Assignment_Week4/VFS.c

```
@@ -0,0 +1,522 @@
+            directory->child = childNode->next;
+        childNode->previous->next = childNode->next;
+        childNode->next->previous = childNode->previous;
+    }
+}
+
+void initializeDisk(int numberOfBlocks)
+{
+    if (numberOfBlocks < 1 || numberOfBlocks > 5000)
+        numberOfBlocks = MAX_BLOCKS;
+    totalBlockCount = numberOfBlocks;
+    diskMemory = malloc(BLOCK_SIZE * totalBlockCount);
+    for (int i = 0; i < totalBlockCount; i++)
+        addFreeBlock(i);
+    rootDirectory = createFileNode("/", 1, NULL);
+    currentDirectory = rootDirectory;
+}
+
+void clearFileTree(FileNode *node)
+{
+    if (node == NULL)
+        return;
+    if (node->child)
+    {
+        FileNode *child = node->child;
+        FileNode *iterator = child;
+        do
+        {
+            FileNode *nextChild = iterator->next;
+            clearFileTree(iterator);
+            iterator = nextChild;
+        } while (iterator != child);
+    }
+    if (!node->isDirectory && node->blockList)
+        free(node->blockList);
+    free(node);
+}
+
+void cleanupFileSystem()
+{
+    clearFileTree(rootDirectory);
+    while (freeBlockHead)
+    {
+        FreeBlock *temp = freeBlockHead;
+        freeBlockHead = freeBlockHead->next;
+        free(temp);
+    }
+    if (diskMemory)
+        free(diskMemory);
+}
+
+void createDirectory(char *name)
+{
+    if (!name || getLength(name) == 0)
+    {
+        printf("Invalid name\n");
+        return;
+    }
+    if (findChild(currentDirectory, name))
+    {
+        printf("Directory already exists\n");
+        return;
+    }
+    FileNode *newDirectory = createFileNode(name, 1, currentDirectory);
+    addChild(currentDirectory, newDirectory);
+    printf("Directory '%s' created successfully.\n", name);
+}
+
+void createFile(char *name)
+{
+    if (name == NULL || getLength(name) == 0)
+    {
+        printf("Invalid name\n");
+        return;
+    }
+    if (findChild(currentDirectory, name))
```

There's no differentiation between files and directories, which means that a file and directory could have same name. You should prevent duplicate names within same directory.

---

I have a doubt regarding this review comment and want to clarify the expectation before fixing this.

Example:
mkdir demo
cd demo
mkdir demo
create demo.txt

The current code support this. But if I try to create another file or directory with the same name in the same directory it throws an error. 

So my question is, do we want to enforce global uniqueness across the entire filesystem?

For example: 
mkdir demo
cd demo
mkdir demo

So now the parent is also demo and the child directory is also demo it should throw an error?

---

The assignment specifies that two files or directories in the same directory cannot have the same name, but it does not mention enforcing global uniqueness across the entire filesystem.

Therefore, the current implementation — which ensures uniqueness within the same directory — already meets the requirement.

This check for global uniqueness is not strictly needed, however, it’s good practice to consider including it for robustness and consistency, especially if the filesystem structure is extended later. You can skip this for now and resolve the comment

---

Okay. Noted will keep this in mind. Thank you1